### PR TITLE
RD2: Site Guest User RD Account Access

### LINKED ISSUE #6000: Enhanced Recurring Donations: Installment Opportunity not created when a Site/Community Guest User creates the Recurring Donation, even in System Mode - Without Sharing, and Recurring Donation Name Format potentially not correctly applied
### Summary of the Issue

Winter ’21 enforced security policies which locked down the Site/Community Guest User. See release notes: https://releasenotes.docs.salesforce.com/en-us/winter21/release-notes/rn_networks_guest_user.htm

As a result any Enhanced Recurring Donations enabled Org that is using a site or community guest user to create recurring donations will notice the installment opportunity and associated related records (e.g payment and GAU Allocations) will not be created successfully and will instead generate an NPSP error log entry.

Customers may also notice any recurring donation naming they have applied in NPSP settings is not applied to the recurring donation record correctly, unless they have defined Account/Contact Site Guest User Sharing Rules.

**Additional Background Notes**
- The missing installment opportunity and related records (e.g payment and GAU Allocation if relevant) will be created by the nightly NPSP 06 - Enhanced Recurring Donation Updates scheduled job, or if a user manually runs the Recurring Donations Batch in NPSP Settings. 
- Under Legacy Recurring Donations a Guest User created Recurring Donation does create Installment Opportunities successfully, though Recurring Donation Naming may not be applied correctly if Guest User Sharing Rules for Account/Contacts have not been created

### Steps to Replicate

- Create a single screen flow that gathers Recurring Donation fields and creates a Recurring Donation record using fields from the screen, ensure the Flow is running in System Context Without Sharing, embed the Flow in a Visualforce page and make the Visualforce page the Active Site Home Page.
OR
- Create a Visualforce page and associated Apex Class, declared Without Sharing, and make this the Active Site Home Page
OR
- Create a Public Community, exposing the Contact, Account, Recurring Donation and Opportunity objects.

### Test to Replicate the Issue

**Test Scenario 1: Flow/Visualforce both disables the First Installment and sets the Recurring Donation Name manually**

Flow/Visualforce Page: Sets the npsp__DisableFirstInstallment__c to TRUE upon creation
Flow/Visualforce Page: Sets the Recurring Donation Name upon creation
NPSP Settings: Installment Opportunity Auto-Creation = Always Create Next Installment
NPSP Settings: Recurring Donation Name Format = Disable Auto Naming
Sharing Rules: No Guest User Sharing Rules defined

Results
- Recurring Donation Created and associated to the Contact and Account: Pass
- Installment Opportunity Created: NA
- Recurring Donation Name Format Applied: NA
- Sample Error: No error / NA

**Test Scenario 2: Flow/Visualforce disables the First Installment and NPSP sets the Recurring Donation Name Format**

Flow/Visualforce Page: Sets the npsp__DisableFirstInstallment__c to TRUE upon creation
NPSP Settings: Installment Opportunity Auto-Creation = Always Create Next Installment
NPSP Settings: Recurring Donation Name Format = {!npe03__Donor_Name__c} {!npe03__Amount__c} - Recurring
Sharing Rules: No Guest User Sharing Rules defined

Results
- The Recurring Donation Name does not include the donor’s name. 
- Recurring Donation Created and associated to the Contact and Account: Pass
- Installment Opportunity Created: NA
- Recurring Donation Name Format Applied: Fail (format set as $2 - Recurring i.e. contact name merge is missing)
- Sample Error: No error

**Test Scenario 3: NPSP creates the First Installment and sets the Recurring Donation Name Format**

NPSP Settings: Installment Opportunity Auto-Creation = Always Create Next Installment
NPSP Settings: Recurring Donation Name Format = {!npe03__Donor_Name__c} {!npe03__Amount__c} - Recurring
Sharing Rules: No Guest User Sharing Rules defined

Results
- The Recurring Donation Name does not include the donor’s name and the Installment Opportunity is not created. 
- Recurring Donation Created and associated to the Contact and Account: Pass
- Installment Opportunity Created: Fail
- Recurring Donation Name Format Applied: Fail (format set as $2 - Recurring i.e. contact name merge is missing)
- Sample Error: Object Type = Opportunity; insufficient access rights on cross-reference id: 0012D00000QRSA9 (where the ID is the Household Account ID)

**Test Scenario 4: As per test scenario 3 but with a Guest User Sharing Rule for Accounts**

Organization Wide Defaults: Contact is Controlled By Parent
Account Guest Sharing Rule to share All Account records with the Guest User
Account: Account.TextName NOT EQUAL TO
NPSP Settings: Installment Opportunity Auto-Creation = Always Create Next Installment
NPSP Settings: Recurring Donation Name Format = {!npe03__Donor_Name__c} {!npe03__Amount__c} - Recurring

Results
- The Recurring Donation Name is applied correctly but the Installment Opportunity is not created. 
- Recurring Donation Created and associated to the Contact and Account: Pass
- Installment Opportunity Created: Fail
- Recurring Donation Name Format Applied: Pass
- Sample Error: Object Type = Opportunity; insufficient access rights on cross-reference id: a092D000003MSpy (where the ID is the Recurring Donation ID)

**Test Scenario 5: As per test scenario 3 but with Guest Sharing Rules for Accounts and Recurring Donations**

Account Guest Sharing Rule to share All Account records with the Guest User
Account: Account.TextName NOT EQUAL TO
Recurring Donation Site Guest Sharing Rule to share All Recurring Donation records with the Guest User
Recurring Donation: Recurring Donation Name NOT EQUAL TO
NPSP Settings: Installment Opportunity Auto-Creation = Always Create Next Installment
NPSP Settings: Recurring Donation Name Format = {!npe03__Donor_Name__c} {!npe03__Amount__c} - Recurring

Results
- The Recurring Donation Name is applied correctly but the Installment Opportunity is not created. 
- Recurring Donation Created and associated to the Contact and Account: Pass
- Installment Opportunity Created: Fail
- Recurring Donation Name Format Applied: Pass
- Sample Error: Object Type = Opportunity; insufficient access rights on cross-reference id: a092D000003MSq3 (where the ID is the Recurring Donation ID)

**Test Scenario 6: As per test scenario 3 but with Guest User Sharing Rules for Accounts, Recurring Donations and Opportunities**

Account Guest Sharing Rule to share All Account records with the Guest User
Account: Account.TextName NOT EQUAL TO
Recurring Donation Guest Sharing Rule to share All Recurring Donation records with the Guest User
Recurring Donation: Recurring Donation Name NOT EQUAL TO
Opportunity Guest Sharing Rule to share All Opportunity records with the Guest User
Opportunity: Opportunity Name NOT EQUAL TO
NPSP Settings: Installment Opportunity Auto-Creation = Always Create Next Installment
NPSP Settings: Recurring Donation Name Format = {!npe03__Donor_Name__c} {!npe03__Amount__c} - Recurring

Results
- The Recurring Donation Name is applied correctly but the Installment Opportunity is not created. 
- Recurring Donation Created and associated to the Contact and Account: Pass
- Installment Opportunity Created: Fail
- Recurring Donation Name Format Applied: Pass
- Sample Error: Object Type = Opportunity; insufficient access rights on cross-reference id: a092D000003MSqU (where the ID is the Recurring Donation ID)

### PR DESCRIPTION
**Pull request by andrewyu-salesforce**
As an integration user I want a Recurring Donation to be created by a Site Guest User without error.

# Critical Changes

# Changes

# Issues Closed
- Fixes #6000 

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by andrewyu-salesforce** [comment]
Attach To: [W-8287986](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008hZbKIAU/view)

**Comment by andrewyu-salesforce** [comment]
@RadSFDeveloper @npsp-reedestockton The PR is ready for review

File: src/classes/RD2_Donor.cls [hunk 1]
```diff
@@ -76,8 +76,9 @@ public with sharing class RD2_Donor {
 
     /***
     * @description Gateway to retrieve donor information
+    *   Use without sharing context so Guest Site User can retrieve Account infos
     */
-    public class Gateway implements IGateway {
+    public without sharing class Gateway implements IGateway {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_Donor.cls:81]: Should `RD2_Donor` and `RD2_NamingService` have `inherited sharing`?

**Review by RadSFDeveloper — APPROVED** [review]
Looks good, thank you @andrewyu-salesforce !

**Review by npsp-reedestockton — APPROVED** [review]
Looks good, Andrew.

**Pull request merged**